### PR TITLE
Copy injector code from rbp and expose api

### PR DESCRIPTION
**Pull request by BenLorantfy**
### In this PR:
- [x] Copy code and tests from rbp to this library
- [x] Create API: `createInjectorsEnhancer` (new), `injectSaga`, `useInjectSaga`, `injectReducer`, `useInjectReducer`, `forceReducerReload` (new), `SagaInjectionModes`
- [x] Setup jest
- [x] Setup babel
- [x] Committed `package-lock.json` file. (This is where most of the "lines added" is coming from)
- [x] Ensure lodash imports aren't importing all of lodash

### Still left TODO (In subsequent PRs):
- [ ] Setup linting
- [ ] Setup CI
- [ ] Add docs
- [ ] Setup cjs, es, and umd builds via [bili](https://github.com/egoist/bili)
- [ ] Create a PR into rbp removing injector code and setting up rbp to use this library

If I should do any of the above in this PR let me know, I just thought to keep this PR small-ish I would do these in subsequent ones.

#### Pinning Dependencies in `package.json`
I don't know if this is a good idea for a library.  I've been working recently on reducing my webpack bundle size, and I've noticed that pinned dependencies are more prone to being duplicated. 

For example, if I pin "lodash" to `4.17.15` and a consumer of this library has a "lodash" version of `^4.20.0`, this will cause two versions of "lodash" to be present in the bundle. If I had however specified `^4.17.15`, then one version would be present in the bundle.

#### checkStore
This has been removed from the API.

### CODE REVIEW

**Comment by bwhitty** [comment]
Hey @BenLorantfy fyi we've basically copy-pasted this code into a large monorepo setup and have validated that at least the raw `injectReducer()` and `injectSaga()` APIs work well!

Anything you need to help get this over the line?

**Comment by BenLorantfy** [comment]
Just waiting for a review

**Comment by BenLorantfy** [comment]
@justingreenberg @julienben @gretzky Does anyone have a chance to review this?

File: babel.config.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      ],
+    },
+    test: {
+      plugins: [
+        '@babel/plugin-transform-modules-commonjs',
+        'dynamic-import-node',
+      ],
+    },
+  },
+};
```
Comment (julienben) [inline babel.config.js]: A few of the plugins in this file are irrelevant for this library no?

I would assume these don't do anything here:
- dynamic import
- react-inline-elements and react-constant-elements (they're only for JSX AFAIK)

Comment (julienben) [reply babel.config.js]: Also, class-properties.

File: jest.config.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+  // A list of paths to directories that Jest should use to search for files in
+  // roots: [
+  //   "<rootDir>"
+  // ],
+
+  // Allows you to use a custom runner instead of Jest's default test runner
+  // runner: "jest-runner",
+
+  // The paths to modules that run some code to configure or set up the testing environment before each test
+  setupFiles: ["@babel/polyfill"],
```
Comment (julienben) [inline jest.config.js]: @babel/polyfill has been deprecated in favor of core-js v3 I think. react-boilerplate uses the latter (via react-app-polyfill but you can do core-js directly here).

Comment (julienben) [reply jest.config.js]: Also, maybe remove all the irrelevant comments in this config file? It would be more readable.

File: package.json [hunk 3]
```diff
@@ -0,0 +1,44 @@
+{
+  "name": "react-redux-injectors",
+  "version": "0.1.0",
+  "main": "lib/index.js",
```
Comment (julienben) [inline package.json]: This should be dist/index.js IMO. Unless you're waiting til after you do the build/publish mechanism?

File: package.json [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    "jest-cli": "^24.8.0",
+    "react": "^16.5.0",
+    "react-dom": "^16.5.0",
+    "react-redux": "^7.1.0",
+    "react-test-renderer": "^16.8.6",
+    "redux-saga": "^1.0.5"
+  },
+  "peerDependencies": {
+    "react": "^16.5.0",
+    "react-dom": "^16.5.0",
```
Comment (julienben) [inline package.json]: We have hooks in the library so these should be 16.8 no?

File: src/checkStore.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import { conformsTo, isFunction, isObject } from 'lodash';
```
Comment (julienben) [inline src/checkStore.js]: You can take care of one your TODOs fairly easily here:
```js
import conformsTo from 'lodash/conformsTo';
import isFunction from 'lodash/isFunction';
import isObject from 'lodash/isObject';
```

File: src/injectSaga.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    static displayName = `withSaga(${WrappedComponent.displayName ||
+      WrappedComponent.name ||
+      'Component'})`;
+
+    constructor(props, context) {
+      super(props, context);
+
+      this.injectors = getInjectors(context.store);
+
+      this.injectors.injectSaga(key, { saga, mode }, this.props);
```
Comment (julienben) [inline src/injectSaga.js]: Here the injectSaga function receives the WrappedComponent's props. That's how it is in react-boilerplate as well but:
- It's not documented anywhere
- It eventually reaches runSaga as an arg but it's not clear what you'd do with props there
- I'm almost certain nobody uses it
- The useInjectSaga hook doesn't do it

I think this is a good chance to get rid of this discrepancy between the HOC and the Hook so my advice is to remove it. If we missed the reason for its existence, somebody will file an issue.

File: package.json [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    "@babel/polyfill": "^7.4.4",
+    "@babel/preset-env": "^7.5.5",
+    "@babel/preset-react": "^7.0.0",
+    "@testing-library/react": "^8.0.6",
+    "babel-eslint": "^10.0.2",
+    "jest-cli": "^24.8.0",
+    "react": "^16.5.0",
+    "react-dom": "^16.5.0",
+    "react-redux": "^7.1.0",
+    "react-test-renderer": "^16.8.6",
```
Comment (julienben) [inline package.json]: Aren't you supposed to have the same version for react, react-dom and react-test-renderer? Might want to upgrade to 16.9.0 while you're at it.

File: package.json (hunk 4 above)
Comment (julienben) [reply package.json]: (IMO, it can stay as 16.5 in peer deps only if the HOC API works despite hooks not being available in React < 16.8.)

Comment (julienben) [reply package.json]: Actually, contextType was added in 16.6 so it should be at least 16.6 if not 16.8.

File: package.json [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  "main": "lib/index.js",
+  "types": "index.d.ts",
+  "repository": "git@github.com:react-boilerplate/injectors.git",
+  "author": "react-boilerplate",
+  "license": "MIT",
+  "scripts": {
+    "build": "babel src --out-dir lib",
+    "test": "jest"
+  },
+  "devDependencies": {
```
Comment (julienben) [inline package.json]: Good point about bundle size. Can we pin dev deps though?

File: src/checkStore.js (hunk 5 above)
Comment (julienben) [reply src/checkStore.js]: You can then enforce this with [this ESLint rule](https://github.com/wix/eslint-plugin-lodash/blob/master/docs/rules/import-scope.md).

File: babel.config.js (hunk 1 above)
Comment (julienben) [reply babel.config.js]: 👌

File: package.json [hunk 9]
```diff
@@ -11,28 +11,28 @@
...
+    "@testing-library/react": "8.0.9",
+    "babel-eslint": "10.0.3",
+    "core-js": "3.2.1",
+    "jest-cli": "24.9.0",
+    "react": "16.9.0",
+    "react-dom": "16.9.0",
+    "react-redux": "7.1.1",
+    "react-test-renderer": "16.9.0",
+    "redux-saga": "1.0.5",
+    "regenerator-runtime": "0.13.3"
```
Comment (julienben) [inline package.json]: Do we need regenerator runtime? For the test which imports redux-saga?

File: package.json (hunk 4 above)
Comment (julienben) [reply package.json]: Just curious. Say a dev is on 16.6 and uses the HOC. Obviously they can't use the hooks... But does the fact that our code imports hooks from React break the entire library for them or they can still use the HOC successfully?

File: package.json (hunk 9 above)
Comment (julienben) [reply package.json]: 👍

File: babel.config.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    production: {
+      plugins: [
+        '@babel/plugin-transform-react-inline-elements',
+        '@babel/plugin-transform-react-constant-elements',
+      ],
+    },
+    test: {
+      plugins: [
+        '@babel/plugin-transform-modules-commonjs',
+        'dynamic-import-node',
```
Comment (julienben) [inline babel.config.js]: This is necessary for Jest? Sorry for being nitpicky but this is a new/clean setup so I'm trying to learn from this and see how we might also potentially clean up some stuff in react-boilerpate.

File: package.json (hunk 4 above)
Comment (julienben) [reply package.json]: Thanks for checking!

File: babel.config.js (hunk 10 above)
Comment (julienben) [reply babel.config.js]: Well while I'm being nitpicky, maybe remove it from package.json and package-lock.json too no?

Comment (julienben) [reply babel.config.js]: Ah yes maybe it was never in package.json but we were able to use it with babel because it's used by one of our deps.

**Pull request merged**